include/revng/Model/Filters.h

```diff
@@ -46,11 +50,39 @@ auto unwrap(const Type &Value, CallableType &&Callable) {
 
 } // namespace detail
 
+//
+// Generic filters
+//
+
+constexpr auto DereferencePointers = std::views::filter([](auto *Ptr) {
+                                       return Ptr != nullptr;
+                                     })
+                                     | std::views::transform([](auto *T) {
+                                         return *T;
+                                       });
+
+//
+// Type level filters
+//
+
 constexpr auto Scalar = std::views::filter([](const auto &Value) {
   return detail::unwrap(Value, [](auto &&V) { return V.isScalar(); });
 });
 constexpr auto NonScalar = std::views::filter([](const auto &Value) {
   return detail::unwrap(Value, [](auto &&V) { return not V.isScalar(); });
 });
 
+//
+// Type definition level filters
+//
+
+constexpr auto Struct = std::views::transform([](auto &&T) {
+                          return llvm::dyn_cast<model::StructDefinition>(&*T);
+                        })
+                        | DereferencePointers;
+constexpr auto Union = std::views::transform([](auto &&T) {
+                         return llvm::dyn_cast<model::UnionDefinition>(&*T);
+                       })
+                       | DereferencePointers;
```

Comment:
`FilterStruct` and `FilterUnion`? It's unclear that you're silently ignoring stuff.

Reply:
They are already named `model::filter::Struct` and `model::filter::Union`.

Reply:
Aaaahn